### PR TITLE
Fixing #2302

### PR DESCRIPTION
**Pull request by stevekogo**

### CODE REVIEW

**Comment by stevekogo** [comment]
@lilyclements Can you please help review this pull request.

File: instat/dlgCorrelation.vb [hunk 1]
```diff
@@ -34,16 +34,15 @@ Public Class dlgCorrelation
         ucrReceiverFirstColumn.Selector = ucrSelectorCorrelation
         ucrReceiverSecondColumn.Selector = ucrSelectorCorrelation
         ucrReceiverMultipleColumns.Selector = ucrSelectorCorrelation
-        ucrReceiverFirstColumn.SetDataType("numeric")
-        ucrReceiverSecondColumn.SetDataType("numeric")
+        ucrReceiverFirstColumn.SetIncludedDataTypes({"numeric"})
+        ucrReceiverSecondColumn.SetIncludedDataTypes({"numeric"})
```
Comment (lilyclements) [inline instat/dlgCorrelation.vb]: Out of interest, why have you changed this from "SetDataType" to "SetIncludedDataTypes"? I always thought you would use the latter if you had more than one data type to set. Is there a difference I wasn't aware of?

File: instat/dlgCorrelation.vb [hunk 2]
```diff
@@ -196,4 +196,18 @@ Public Class dlgCorrelation
             ucrBase.OKEnabled(False)
         End If
     End Sub
+
+    Private Sub ucrSelectorCorrelation_DataFrameChanged() Handles ucrSelectorCorrelation.DataFrameChanged
+        Pairs()
+    End Sub
+
+    Public Sub Pairs()
+        Dim clsTempFunc As RFunction
```
Comment (lilyclements) [inline instat/dlgCorrelation.vb]: Which bug in ggpairs is this trying to fix?

Comment (lilyclements) [inline instat/dlgCorrelation.vb]: Additionally, do we need to declare clsTempFunc here as well as at the start of the class? I thought we could just declare it the once?

**Comment by lilyclements** [comment]
This still seems to have one of the bugs discussed in #2302 which is that if multiple check boxes are selected in the subdialog, only one is run.

PairwisePlot, CorrPlot and ScatterplotMatrix bring up plots. At this stage we would only expect to get one graph up since that needs to be sorted still, so it is not useful to test with that.
Correlation Matrix brings up the corr matrix in the output window. However, this currently works when only the correlation matrix is selected. We want this matrix to be shown regardless of what else is selected.

We want to have that whatever in the check boxes are selected run. Currently, we can see this is not the case because if I select one of the plots and the correlation matrix, the plot runs but the matrix does not.

File: instat/dlgCorrelation.vb (hunk 2 above)
Comment (stevekogo) [inline instat/dlgCorrelation.vb]: The bug we are fixing here is explained in #2302. I will add comments on the code to explain how this is addressed.

**Comment by stevekogo** [comment]
@lilyclements Thanks for the wonderful comments, let me address them then update the pull request.

**Comment by stevekogo** [comment]
@lilyclements , @dannyparsons  I have scrutinized the best way to display correlation matrix and graphs in correlation dialogue and I have found out two different implementations with their pros and cons

1. We may choose to display whatever is checked even if all the three graphs are but the save graph option is compromised since we cannot specify which graph to save and which ones to just display.

2. We may choose to display one graph with or without the correlation matrix at a time. This works well with the save graph. In this case, we may chose to restrict graph checkboxes to only one or use radio buttons.

I chose to implement the second option(I am yet to restrict the checkboxes to one graph)but your view on this is highly welcome.

**Comment by lilyclements** [comment]
I feel 1. could be beneficial and we could have three "Save Graph" options - one in the "Pairwise Plot" tab, one in the "Correlation Plot" tab and the final in the "Scatterplot" tab as it does not restrict the user as much?

Here's another bug I've just found on this one. It is that
I have two datasets. I open "Correlation" and put in dataset A, then I select "Options" and open the sub dialog. I go to "Scatterplot Matrix". The dataset available for factors is from dataset A.
Then I go to my main dialog, change it to dataset B and the dataset available for factors under "Scatterplot Matrix" is still from dataset A

**Comment by stevekogo** [comment]
@lily The bug in the scatterplot dataset is now fixed, you may carry on with the review.

As with the graphs check boxes @rstern may advise us on the best way to implement this. Here is my explicit analogy on the matter.

If we choose to allow all the graphs check boxes, then saving option is compromised but @lilyclements was suggesting we have three save graph options on the sub dialog, one for each graph. The implementation of this will not be easy since one graph command is set as the base function upon click OK but  @dannyparsons may advise on the ease of doing this.

If we choose to display one graph at a time, we may keep the check boxes but a neater implementation will be radio buttons. With this implementation, we just need one save graph option and it is easy to implement.

Your advice on the right implementation is highly welcome.

**Comment by rdstern** [comment]
I would be happy with one graph at a time.  I am more interested in the options we can suggest for each of the different graph types.  I would like users to be able to get the graph type of their choice as they would really like to see it.  I am not convinced about multiple windows.

Later, if there are times when we want more than one graph type, then perhaps this could be a combine graph - so it is again a single screen that is produced.  That could be just to show users what is possible, and then they use the type of their choice.

Where we can have more is in the help system.

Roger

**Comment by lilyclements** [comment]
Okay great, so the graphs will be radio buttons then? Or at least check boxes where only one can be selected.
How about the correlation matrix? Can that be selected with a graph?

**Comment by lilyclements** [comment]
After a discussion yesterday, I think it was decided that we should not have "Correlation Matrix" in the sub dialog, and instead this should be a feature of the main dialog for when "Multiple" is selected.
We can have an option on the main dialog to display this matrix output or not, however, it will always be run.
We can then change "Options" to "Graph Options" and the sub dialog can contain graphs.

@rdstern @dannyparsons - is that correct? Have I missed anything out?

**Comment by rdstern** [comment]
That's right.  I am also relaxed on whether a single graph option is chosen, or they can all be selected.  We also agreed that it would be possible for there to be no printout at all, i.e. if you didn't select a graph and opted not to have the correlation matrix displayed.  This could still be sensible, particularly if you chose to save the correlation object.  (For example we have our general dialogues for operations on objects.  I am sure print works, so you could always choose to display the correlations in the future.)

**Comment by lilyclements** [comment]
@stevekogo how have you been getting on with this?

**Comment by rstern** [comment]
Just a heads up, you've been accidentally including me in your discourses. I suspect you were looking to add rdstern, a tag I noticed in one of the notification emails. My account is rstern (no 'd').
I unsubscribed. Good luck with your project!

**Comment by stevekogo** [comment]
@africanmathsinitiative/instat This is ready for review.

File: instat/sdgCorrPlot.vb [hunk 3]
```diff
@@ -53,42 +55,43 @@ Public Class sdgCorrPlot
...
         ucrSaveGraph.SetDataFrameSelector(dlgCorrelation.ucrSelectorCorrelation.ucrAvailableDataFrames)
         ucrSaveGraph.strPrefix = "CorGraph"
-        chkCorrelationMatrix.Checked = True
-        chkPairwisePlot.Checked = False
-        chkCorrelationPlot.Checked = False
-        chkScatterplotMatrix.Checked = False
+        dlgCorrelation.chkCorrelationMatrix.Checked = True
+        rdoNone.Checked = True
+        tbSaveGraphs.Visible = False
         nudAlphaScatter.Value = 1
```
Comment (lilyclements) [inline instat/sdgCorrPlot.vb]: We also want
nudAlphaCor.Value = 1 here as well as
chkLabel to be checked and ucrReceiveFactor to be reset

File: instat/dlgCorrelation.vb [hunk 4]
```diff
@@ -181,13 +192,17 @@ Public Class dlgCorrelation
...
+            ucrBase.clsRsyntax.RemoveAssignTo()
+            If (Not ucrReceiverFirstColumn.IsEmpty()) AndAlso (Not ucrReceiverSecondColumn.IsEmpty()) AndAlso (rdoPearson.Checked = True Or rdoKendall.Checked = True Or rdoSpearman.Checked = True) Then
                 ucrBase.OKEnabled(True)
             Else
                 ucrBase.OKEnabled(False)
             End If
-        ElseIf (rdoMultipleColumns.Checked = True) And (sdgCorrPlot.chkCorrelationMatrix.Checked OrElse sdgCorrPlot.chkCorrelationPlot.Checked OrElse sdgCorrPlot.chkScatterplotMatrix.Checked OrElse sdgCorrPlot.chkPairwisePlot.Checked) Then
-            If (Not ucrReceiverMultipleColumns.IsEmpty()) And ucrReceiverMultipleColumns.lstSelectedVariables.Items.Count > 1 And (rdoCompleteRowsOnly.Checked = True Or rdoPairwise.Checked = True) AndAlso (rdoPearson.Checked = True Or rdoKendall.Checked = True Or rdoSpearman.Checked = True) Then
+        ElseIf (rdoMultipleColumns.Checked = True) AndAlso (ucrSaveModel.chkSaveModel.Checked OrElse chkCorrelationMatrix.Checked OrElse sdgCorrPlot.rdoCorrelationPlot.Checked OrElse sdgCorrPlot.rdoScatterplotMatrix.Checked OrElse sdgCorrPlot.rdoPairwisePlot.Checked) Then
```
Comment (lilyclements) [inline instat/dlgCorrelation.vb]: Currently, if I have "None" selected in the sub dialog and uncheck the correlation matrix box on the main dialog, it runs first time, and then when reopening the dialog "OK" is no longer enabled.
We want "OK" to be enabled no matter what is checked in the sub dialog, and we want "OK" to be enabled even if the correlation matrix is unchecked.

If the correlation matrix is unchecked we do not want the output to be displayed but still want the matrix to run and be saved

**Pull request merged**